api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java

```diff
@@ -269,11 +269,16 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
 
     @Override
     public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
-      return ROWS_MIGHT_MATCH;
+      // in(col, {X, Y}) => eq(col, x) OR eq(col, x)
+      if (literalSet.stream().anyMatch(v -> eq(ref, toLiteral(v)))) {
```

Comment:
This should not convert each value to a literal on every invocation. Instead, this PR can add a `literals` method to `BoundSetPredicate` to access the values as literals. That way, the conversion happens once and can be lazy.

Reply:
It looks like this probably just needs the `Comparator`, not the literals.

Reply:
Thanks for the comments and update the code accordingly.